core/services/pipeline/orm.go

```diff
@@ -0,0 +1,194 @@
+		err = tx.Exec(`
+            INSERT INTO pipeline_task_runs (
+                pipeline_run_id, pipeline_task_spec_id, created_at
+            )
+            SELECT ? AS pipeline_run_id, id AS pipeline_task_spec_id, NOW() AS created_at
+            FROM pipeline_task_specs
+            WHERE pipeline_spec_id = ?
+        `, run.ID, run.SpecID).Error
+		return errors.Wrap(err, "could not create pipeline task runs")
+	})
+	return runID, err
+}
+
+// WithNextUnclaimedTaskRun chooses any arbitrary incomplete TaskRun from the DB
+// whose parent TaskRuns have already been processed.
+func (o *orm) WithNextUnclaimedTaskRun(fn func(ptRun TaskRun, predecessors []TaskRun) Result) error {
+	return ormpkg.GormTransaction(o.db, func(tx *gorm.DB) error {
+		var ptRun TaskRun
+		var predecessors []TaskRun
+
+		// NOTE: This could conceivably be done in pure SQL and made marginally more efficient by preloading with joins
+
+		// Find the next unlocked, unfinished pipeline_task_run with no uncompleted predecessors
+		err := tx.Table("pipeline_task_runs AS successor").
+			Set("gorm:query_option", "FOR UPDATE OF successor SKIP LOCKED"). // Row-level lock
+			Joins("LEFT JOIN pipeline_task_runs AS predecessors ON successor.id = predecessor.successor_id AND predecessors.finished_at IS NULL").
+			Where("predecessors.id IS NULL").
+			Where("successor.finished_at IS NULL").
+			Preload("TaskSpec").
+			Order("id ASC").
+			First(&ptRun).
+			Error
+		if err != nil {
+			return errors.Wrap(err, "error finding next task run")
+		}
+
+		// Find all the predecessors
+		err = tx.Where("successor_id = ?", ptRun.ID).Find(&predecessors).Error
+		if err != nil {
+			return errors.Wrap(err, "error finding task run predecessors")
+		}
+
+		// Call the callback and convert its output to a format appropriate for the DB
+		result := fn(ptRun, predecessors)
+		var out *JSONSerializable
+		var errString null.String
+		if result.Value != nil {
+			out = &JSONSerializable{Value: result.Value}
+		} else if result.Error != nil {
+			errString = null.StringFrom(err.Error())
+		}
+
+		// Update the task run record with the output and error
+		err = tx.Exec(`
+            UPDATE pipeline_task_runs
+            SET output = ?, error = ?, finished_at = ?
+            WHERE id = ?
+        `, out, errString, time.Now(), ptRun.ID).Error
+
+		return errors.Wrap(err, "could not mark pipeline_task_run as finished")
+	})
+}
+
+func (o *orm) AwaitRun(ctx context.Context, runID int64) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		var done bool
+		err := o.db.Exec(`
+            SELECT bool_and(pipeline_task_runs.error IS NOT NULL OR pipeline_task_runs.output IS NOT NULL)
+            FROM pipeline_job_runs
+            JOIN pipeline_task_runs ON pipeline_task_runs.pipeline_job_run_id = pipeline_job_runs.id
+            WHERE pipeline_job_runs.id = $1
+        `, runID).Scan(&done).Error
+		if err != nil {
+			// TODO: log error
```

Comment:
This would indicate something fairly catastrophic. Might be ok to return error here.